### PR TITLE
Adding display transformation base

### PR DESCRIPTION
**Pull request by cam156**
Here is another way to do the Factory.

Not sure if this better or worse than the validation pattern.  The is a single factory for the system, which can have the transformations set on it.

### CODE REVIEW

File: app/cho/display_transformation/factory.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+module DisplayTransformation
+  class Factory
+    def self.transformations=(new_transformations)
+      non_transformations = new_transformations.reject { |_label, transformation| transformation.is_a? DisplayTransformation::Transformation }
```
Comment (awead) [inline app/cho/display_transformation/factory.rb]: Can `_label` be removed?

File: app/cho/display_transformation/factory.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+      if !non_transformations.empty?
```
Comment (awead) [inline app/cho/display_transformation/factory.rb]: `non_transformations.present?` reads better and should work with empty arrays.

File: app/cho/display_transformation/factory.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        error_list = non_transformations.map { |label, _err_validator| label }.join(', ')
```
Comment (awead) [inline app/cho/display_transformation/factory.rb]: Can `_err_validator` be removed here?

File: app/cho/display_transformation/factory.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+        raise "Invalid display transofrmations(s) in transformation list: #{error_list}"
+      end
+
+      @transformations = new_transformations
```
Comment (awead) [inline app/cho/display_transformation/factory.rb]: I like the fact that that the setter is at the end of the method which works as an explicit return.

File: app/cho/display_transformation/factory.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    end
+
+    def self.transformations
+      @transformations ||= { none: DisplayTransformation::None.new }
```
Comment (awead) [inline app/cho/display_transformation/factory.rb]: Here I think this is much better than the Validation class, where it was getting set in parameters. However, we would we have to explicitly add the none transformation to the factory initialization, unless we want to always include none by default. So would something like:
``` ruby
@transformations.merge(none: DisplayTransformation::None.new)
```
 be better so there's always `none` there by default?

File: app/cho/display_transformation/transformation.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+module DisplayTransformation
+  class Transformation
```
Comment (awead) [inline app/cho/display_transformation/transformation.rb]: what about `Base` here? Or is that too overloaded? "Transformation-Transformation" sounds wordy.

File: app/cho/display_transformation/transformation.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
...
+    def transform(_field)
+      raise 'DisplayTransformation.transform is abstract. Children must implement.'
```
Comment (awead) [inline app/cho/display_transformation/transformation.rb]: How do you feel about `NotImplementedError`? Although, I did find this: http://chrisstump.online/2016/03/23/stop-abusing-notimplementederror/

File: spec/cho/display_transformation/factory_spec.rb [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+      let(:transformations) { { mine: MyTransformation.new, abc: '123' } }
+
+      it { within_block_is_expected.to raise_exception 'Invalid display transofrmations(s) in transformation list: abc' }
+    end
+  end
+
+  describe '#tranformations' do
+    subject { described_class.transformations }
+
+    let(:transformations) { {} }
```
Comment (awead) [inline spec/cho/display_transformation/factory_spec.rb]: Shouldn't transformations contain `none`?

File: app/cho/display_transformation/transformation.rb (hunk 6 above)
Comment (awead) [reply app/cho/display_transformation/transformation.rb]: Also Abstract, or AbstractTransformation

File: app/cho/display_transformation/factory.rb (hunk 1 above)
Comment (cam156) [reply app/cho/display_transformation/factory.rb]: @awead since this is a hash I do not believe it can be.

**Pull request merged**